### PR TITLE
fix literal-pow to return the right type when the base is -1

### PR DESCRIPTION
**Pull request by oscardssmith**
fixes https://github.com/JuliaLang/julia/issues/53735.

### CODE REVIEW

File: base/intfuncs.jl [hunk 1]
```diff
@@ -369,8 +369,19 @@ const HWNumber = Union{HWReal, Complex{<:HWReal}, Rational{<:HWReal}}
 @inline literal_pow(::typeof(^), x::HWNumber, ::Val{1}) = x
 @inline literal_pow(::typeof(^), x::HWNumber, ::Val{2}) = x*x
 @inline literal_pow(::typeof(^), x::HWNumber, ::Val{3}) = x*x*x
-@inline literal_pow(::typeof(^), x::HWNumber, ::Val{-1}) = inv(x)
-@inline literal_pow(::typeof(^), x::HWNumber, ::Val{-2}) = (i=inv(x); i*i)
+@inline function literal_pow(::typeof(^), x::HWNumber, ::Val{-1})
+    if x == -one(x)
+        return isodd(p) ? one(x) : -one(x)
```
Comment (fingolfin) [inline base/intfuncs.jl:374]: What is `p`?

Comment (fingolfin) [reply base/intfuncs.jl:374]: Maybe
```suggestion
        return x
```

File: base/intfuncs.jl [hunk 2]
```diff
@@ -369,8 +369,19 @@ const HWNumber = Union{HWReal, Complex{<:HWReal}, Rational{<:HWReal}}
 @inline literal_pow(::typeof(^), x::HWNumber, ::Val{1}) = x
 @inline literal_pow(::typeof(^), x::HWNumber, ::Val{2}) = x*x
 @inline literal_pow(::typeof(^), x::HWNumber, ::Val{3}) = x*x*x
-@inline literal_pow(::typeof(^), x::HWNumber, ::Val{-1}) = inv(x)
-@inline literal_pow(::typeof(^), x::HWNumber, ::Val{-2}) = (i=inv(x); i*i)
+@inline function literal_pow(::typeof(^), x::HWNumber, ::Val{-1})
+    if x == -one(x)
```
Comment (KlausC) [inline base/intfuncs.jl:373]: This change introduces even more type instability, now depending on the _value_ of the `x`  argument.

Other question: why special-case `-1` and not `1`?

File: base/intfuncs.jl [hunk 3]
```diff
@@ -369,8 +369,19 @@ const HWNumber = Union{HWReal, Complex{<:HWReal}, Rational{<:HWReal}}
...
+        return isodd(p) ? one(x) : -one(x)
+    end
+    inv(x)
+end
+@inline function literal_pow(::typeof(^), x::HWNumber, ::Val{-2})
+    if x == -one(x)
```
Comment (KlausC) [inline base/intfuncs.jl:379]: See previous comment on line 373

File: base/intfuncs.jl [hunk 4]
```diff
@@ -379,7 +390,9 @@ const HWNumber = Union{HWReal, Complex{<:HWReal}, Rational{<:HWReal}}
 # for other types, define x^-n as inv(x)^n so that negative literal powers can
 # be computed in a type-stable way even for e.g. integers.
 @inline function literal_pow(f::typeof(^), x, ::Val{p}) where {p}
-    if p < 0
+    if x == -one(x)
```
Comment (KlausC) [inline base/intfuncs.jl:393]: See previous comment on line 373

File: base/intfuncs.jl (hunk 2 above)
Comment (oscardssmith) [reply base/intfuncs.jl:373]: the reason to specialize only on -1 is that inv(x) often has a different type than x, but x*x generally is ok. what is a fade where this version is type unstable? unitful?

Comment (KlausC) [reply base/intfuncs.jl:373]: I meant depending on the first argument: `typeof ((-1)^-1) == Int`, while `typeof(1^-1) == Float64)`

**Comment by KristofferC** [comment]
Ref https://github.com/JuliaLang/julia/issues/53735#issuecomment-2000363348

**Comment by JeffBezanson** [comment]
I'm not sure this needs to be changed. Looks like the cure might be worse than the disease.

**Comment by oscardssmith** [comment]
it's really unfortunate that there has to be type weirdness somewhere. that said given the choices we made with literal pow, I think the current behavior unfortunately is just what we're stick with.

**Pull request closed without merging**